### PR TITLE
Fix a bug so that HttpRequestContext.response.hdrs won't be freed twice.

### PR DESCRIPTION
**Pull request by renkelvin**
Just like `response.body`, we should duplicate `response.hdrs`. Otherwise the single pointer will be freed twice.

This issue was not catched because a simulated response is used in the unit tests instead of the actual response.

@markdroth Please take a look and let me if you have any questions, thanks!

### CODE REVIEW

**Comment by markdroth** [comment]
Can you change the tests such that they will catch this kind of problem?

**Comment by renkelvin** [comment]
To catch this kind of problem, I think we need to file a real http request, which may not be a good idea for unit tests. What do you think? Is there a recommended way to do integration tests in gRPC?

**Comment by markdroth** [comment]
Isn't the test plugging in a fake handler for the HTTP request?  Can you just change that fake handler to check that the request has the expected content?

File: src/core/lib/security/credentials/external/external_account_credentials.cc [hunk 1]
```diff
@@ -167,7 +167,16 @@ void ExternalAccountCredentials::OnExchangeTokenInternal(grpc_error* error) {
   } else {
     if (options_.service_account_impersonation_url.empty()) {
       metadata_req_->response = ctx_->response;
```
Comment (renkelvin) [inline src/core/lib/security/credentials/external/external_account_credentials.cc:169]: The whole response was set so the content is actually correct and we've already check it.
This fix is to have a deep copy of `hdrs` and we need to check if the address has a single reference, which seems hard to do in unit tests.

Comment (renkelvin) [reply src/core/lib/security/credentials/external/external_account_credentials.cc:169]: btw, I'll add a comprehensive integration test once the implementation finalized, and this kind of memory issues can be easily catched.

**Review by markdroth — APPROVED** [review]

**Comment by renkelvin** [comment]
CI is green. Commits squashed. It should be safe to merge this pr. Thanks!

**Comment by renkelvin** [comment]
Hey Mark, is it safe to merge this or should we re-trigger the tests?

**Comment by markdroth** [comment]
Known issues: #23822 #24937

**Pull request merged**